crates/edr_provider/src/requests/eth/blocks.rs

```diff
@@ -19,20 +19,64 @@ pub enum HashOrTransaction {
     Transaction(eth::Transaction),
 }
 
+pub fn handle_get_block_by_hash_request(
+    data: &ProviderData,
+    block_hash: B256,
+    transaction_detail_flag: bool,
+) -> Result<Option<eth::Block<HashOrTransaction>>, ProviderError> {
+    data.block_by_hash(&block_hash)?
+        .map(|block| {
+            let total_difficulty = data.total_difficulty_by_hash(block.hash())?;
+            block_to_rpc_output(data, block, total_difficulty, transaction_detail_flag)
+        })
+        .transpose()
+}
+
 pub fn handle_get_block_by_number_request(
     data: &ProviderData,
     block_spec: PreEip1898BlockSpec,
     transaction_detail_flag: bool,
 ) -> Result<Option<eth::Block<HashOrTransaction>>, ProviderError> {
-    match data.block_by_block_spec(&block_spec.into()) {
+    block_by_number(data, &block_spec.into())?
+        .map(
+            |BlockAndTotalDifficulty {
+                 block,
+                 total_difficulty,
+             }| {
+                block_to_rpc_output(data, block, total_difficulty, transaction_detail_flag)
```

Comment:
The only thing that surprises me is passing the entire `data` to these methods that only format things. I believe the two instances are this, and the one formatting transactions.

Both of them are used to obtain the spec id and decided if the type of transactions should be displayed.

Reply:
Right, thanks, there is no point of passing the entire `data` anymore. Created a follow PR here: https://github.com/NomicFoundation/hardhat/pull/4637